hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -310,7 +310,23 @@ def __init__(self, resp):
 
     async def read(self, n: int = -1) -> bytes:
```

Comment:
OK, sorry about all the back and forth.

I think based on our Zulip thread the right answer is: we need a new method called `readexactly`. It should take one argument, `n` and it should read exactly that many bytes. I think we should also `assert n > 0` because it doesn't make any sense to `readexactly(-1)`.

Reply:
We'll then need to make corresponding changes to the other PR so that where it needs exactly n bytes, it asks for exactly n bytes.